### DISCUSSION TITLE
Where can I change the AST node's value?

### DISCUSSION THREAD

**Discussion by negahama** [eclipse-langium/langium#1641] [Q&A] [answered]
In my DSL
```
def getCharAt(index: number) {
    ...
}
getCharAt(-1)
```
I want to change the argument in `getCharAt()` from `-1` to `0`.
I know that Langium's validator can notify the error and users need to modify for themself.
But This is example for explanation. I have to changed the value by programming at any cost.
So I ask for everyone's help

If I need to change(or append) AST node, Where can I do it? Scope computation? linker? validator? I have no idea at all.
And how can I do?(Can it even possible to change(or append) the AST node?)
At least, could you tell me where can I get the information or samples?

Thanks in advance...

**Comment by cdietrich** [comment]
do you want to change it in the file (refactoring) or in the ast, if in the ast you want to change all -1 to 0, if yes you might look into datatype rules and value converters

**Reply by negahama** [reply]
Thanks a lot...
Your advice is no corresponding to my case, But I didn't know these(datatype rules and value converter) at all..
I think your information will be useful in the future. Thank you so much.

**Reply by cdietrich** [reply]
What is your exact case

**Reply by negahama** [reply]
@cdietrich 
I am developing my new language that is transpiled to TypeScript codes.
My language has a feature that is index numbering of array start from 1 just like Lua script.
So, I have to translate index of array in my language to `original value - 1`.
for example
```
// in my code
var array[1, 2, 3]
var a = array[1] // a = 1

// in translated TypeScript
let array[1, 2, 3]
let a = array[0] // a = 1
```
I implemented this feature(indexing of array by literal)... It's not difficult. 
It's easy to determine array expression(e.g `array[ ]`) and literal in array(`1`) in Langium.

But string's function is not easy
```
var s: string = 'this is sample'
s.charAt(somefunction())

// I have to translate below
let s: string = 'this is sample'
s.charAt(somefunction() - 1)
```

I can find the `charAt` in the Scope computation phase,
But I don't know how to change the `somefunction()`'s value OR to append the new node which is equivalent to `-1` in `charAt`.

I don't know **proper place where it was performed**.
I even don't know **it is possible to change or append the Ast node manually.**

Thank you in advance

**Reply by msujew** [reply, marked as answer]
Hey @negahama,

you shouldn't attempt to change the AST. I would instead recommend you to perform the offset (i.e. `-1`) operation during the transpilation. That way you don't need to keep track of the offset during all other operations (like validation, etc.). You would essentially do something like this:

```ts
function transpileExpression(expression: Expression): string {
  const text = transpileExpressionInternal(expression);
  if (isOffsetRequired(expression)) {
    return `${text} - 1`;
  } else {
    return text;
  }
}

function isOffsetRequired(expression: Expression): boolean {
  const container = expression.$container;
  if (isArrayAccess(container)) {
    return true;
  }
  if (isFunctionCall(container)) {
    const isStringIndexFunction = // compute whether it's a function like `chatAt`
    return isStringIndexFunction;
  }
  return false;
}
```

**Reply by negahama** [reply]
@msujew Thank you so much...
I guess I couldn't change the Ast Node manually. your response make me get a confidence and good example...
Have a good time for this weekend.
